### PR TITLE
CFE-977, CFE-978, CFE-982: Introduce history interface with its Reader and Append implementation

### PR DESCRIPTION
**Pull request by aguidirh**
# Description

This feature creates and read a history file. This file contains all blobs mirrored in a mirrorToDisk flow.

The `Read func` will get the latest history file (format is `.history-time.RFC3339` example: `.history-2023-11-22T11:24:14Z`) if no date is specified. In case a date is specified in the NewHistory func, the Read func will return the latest history before the specified date.

The `Append func` will call the `Read func` to get the blobs from a previous history file and will generate a new history file with the content of the previous history file + the new blobs mirrored in the current run.

Fixes # [CFE-977](https://issues.redhat.com/browse/CFE-977), [CFE-978](https://issues.redhat.com/browse/CFE-978) and [CFE-982](https://issues.redhat.com/browse/CFE-982)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Unit test under `v2/pkg/history/history_test.go`

## Expected Outcome
Unit tests should `pass`

### CODE REVIEW

File: v2/pkg/history/history.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				}
+			}
+		}
+	}
+
+	return latestFile, err
+}
+
+func isHistoryFile(historyFile fs.DirEntry) bool {
+	return !historyFile.IsDir() && strings.Contains(historyFile.Name(), historyNamePrefix)
```
Comment (sherine-k) [inline v2/pkg/history/history.go:101]: or hasPrefix?

File: v2/pkg/history/history.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	"os"
+	"strings"
+	"time"
+
+	clog "github.com/openshift/oc-mirror/v2/pkg/log"
+)
+
+var log clog.PluggableLoggerInterface
+
+type OSFileCreator struct{}
```
Comment (sherine-k) [inline v2/pkg/history/history.go:16]: I like this idea!

File: v2/pkg/history/history_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+type nopCloser struct {
+	io.Writer
+}
+
+func (m MockFileCreator) Create(name string) (io.WriteCloser, error) {
+	m.Buffer = new(bytes.Buffer)
+	return nopCloser{m.Buffer}, nil
+}
```
Comment (sherine-k) [inline v2/pkg/history/history_test.go:24]: I like the idea! no tmp folder, no files on disk!
:+1:

File: v2/pkg/history/history_test.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			history, err := NewHistory(test.workingDir, test.before, clog.New("trace"), MockFileCreator{})
+			assert.NoError(t, err)
+			err = history.Append(test.blobsToAppend)
+
+			if test.expectedError != "" {
+				assert.EqualError(t, err, test.expectedError)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
```
Comment (sherine-k) [inline v2/pkg/history/history_test.go:150]: Any way you could also compare the new history content to expectedHist?

**Review by sherine-k — COMMENTED** [review]
Thanks @aguidirh ! great work!
One small question for integration #734: Can the digests included in .history also contain the algorithm (ex: `sha256:`)? 
The tests don't show it but I don't see why it couldn't.

**Review by lmzuccarelli — COMMENTED** [review]
@aguidirh - agreed great work - nice and clean

**Pull request merged**